File: Shreyansh-Parashar/2DArray/spiralMatrix.java
import java.util.*;
class spiralMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        int r1 = 0;
        int r2 = matrix.length - 1;
        int c1 = 0;
        int c2 = matrix[0].length - 1;
        List<Integer> ans = new ArrayList<>();

        while(r1 <= r2 && c1 <= c2){
            for(int c = c1; c <= c2; c++){
                ans.add(matrix[r1][c]);
            }

            for(int r = r1 + 1; r <= r2; r++){
                ans.add(matrix[r][c2]);
            }

            if(r1 < r2 && c1 < c2){
                for(int c = c2 -1; c >= c1; c--){
                    ans.add(matrix[r2][c]);
                }
                for(int r = r2 - 1; r > r1; r--){
                    ans.add(matrix[r][c1]);
                }
            }
            r1++;
            r2--;
            c1++;
            c2--;
        }
        return ans;
    }
    public static void main(String[] args) {
        spiralMatrix obj = new spiralMatrix();
        Scanner sc = new Scanner(System.in);


        System.out.print("Enter now of rows: ");
        int row = sc.nextInt();

        System.out.print("Enter number of columns: ");
        int col = sc.nextInt();

        int[][] matrix = new int[row][col];

        System.out.println("Enter 2D Matrix: ");
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                matrix[i][j] = sc.nextInt();
            }           
        }
        List<Integer> result = obj.spiralOrder(matrix);
        System.out.println("Spiral Order: " + result);
        sc.close();
        
    }
}

File: Ashutosh-Singh/2D-Array/Spiral-Matrix.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        int row = matrix.size();
        int col = matrix[0].size();
        
        int count = 0;
        int total = row * col;
        
        int startingCol = 0;
        int endingCol = col - 1;
        int startingRow = 0;
        int endingRow = row - 1;
        
        while(count < total){
            // Traverse starting row
            for(int index = startingCol; count < total && index <= endingCol; index++) {
                ans.push_back(matrix[startingRow][index]);
                count++;
            }
            startingRow++;
            
            // Traverse ending column
            for(int index = startingRow; count < total && index <= endingRow; index++) {
                ans.push_back(matrix[index][endingCol]);
                count++;
            }
            endingCol--;
            
            // Traverse ending row
            for(int index = endingCol; count < total && index >= startingCol; index--) {
                ans.push_back(matrix[endingRow][index]);
                count++;
            }
            endingRow--;
            
            // Traverse starting column
            for(int index = endingRow; count < total && index >= startingRow; index--) {
                ans.push_back(matrix[index][startingCol]);
                count++;
            }
            startingCol++;
        }
        return ans;
    }
};

int main() {
    Solution solution;
    
    int n, m;
    
    // Input matrix dimensions
    cout << "Enter number of rows (n): ";
    cin >> n;
    cout << "Enter number of columns (m): ";
    cin >> m;
    
    vector<vector<int>> matrix(n, vector<int>(m));
    
    // Input matrix elements
    cout << "Enter the elements of the matrix:" << endl;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> matrix[i][j];
        }
    }
    
    // Get the spiral order
    vector<int> result = solution.spiralOrder(matrix);

    // Output the result
    cout << "Spiral order: ";
    for (int val : result) {
        cout << val << " ";
    }
    cout << endl;

    return 0;
}
